### PR TITLE
Refactor worker_create_or_replace_object to accept multiple statements

### PR DESCRIPTION
**Pull request by thanodnl**
Refactor worker_create_or_replace_object to accept multiple statements.

This facilitates the reuse of existing TEXT SEARCH CONFIGURATION objects, specifically when users use one of the pre-configured configurations that ship with postgres.

Text search configuration objects require a list of create+alter commands to correctly configure the objects to have the same semantic meanings. Postgres comes with a dozen or so preconfigured objects of these. Without the ability to compare the states against each other we would re-create all the objects on first use.

Even worse is that we would keep the original configurations as well, under a backup name. This would seriously pollute the
catalogs on the workers, which would get an issue with MX becoming more common.

### CODE REVIEW

File: src/backend/distributed/worker/worker_create_or_replace.c [hunk 1]
```diff
@@ -75,32 +152,69 @@ WrapCreateOrReplace(const char *sql)
 Datum
 worker_create_or_replace_object(PG_FUNCTION_ARGS)
 {
-	text *sqlStatementText = PG_GETARG_TEXT_P(0);
-	const char *sqlStatement = text_to_cstring(sqlStatementText);
-	Node *parseTree = ParseTreeNode(sqlStatement);
+	List *sqlStatements = NIL;
```
Comment (onderkalaci) [inline src/backend/distributed/worker/worker_create_or_replace.c:155]: can we always pass array instead of maintaining both? I guess that would break rolling upgrades?

**Review by onderkalaci — COMMENTED** [review]
As I also noted in the other PR, I think this approach is fine, I'm fine to review this along with the https://github.com/citusdata/citus/pull/5685.

(Though, not reviewed this on yet, just noting my high level take)

File: src/backend/distributed/worker/worker_create_or_replace.c (hunk 1 above)
Comment (thanodnl) [reply src/backend/distributed/worker/worker_create_or_replace.c:155]: It is quite an easy refactor to always require an array to be passed. However we mostly don't change the `worker_*` functions signatures much. Guess it would be ok for Citus 11 to require a full cluster upgrade though, but I didn't want to really think much about what could go wrong changing this function, and instead make it work in a backwards compatible manner.

Let me know what your preference is here, I can easily change it to _only_ take arrays.

File: src/backend/distributed/worker/worker_create_or_replace.c [hunk 2]
```diff
@@ -75,32 +152,69 @@ WrapCreateOrReplace(const char *sql)
...
-	 * since going to the drop statement might require some resolving we will do a check
-	 * if the type actually exists instead of adding the IF EXISTS keyword to the
-	 * statement.
+	 * The first argument is either a text or a list of text's, we normalize this to a
+	 * list of statements for further use, where the variant with just a single text
+	 * argument will be treated as a list of statements with length 1
 	 */
-	ObjectAddress address = GetObjectAddressFromParseTree(parseTree, true);
-	if (ObjectExists(&address))
+	Oid argTypeOid = PG_GETARG_TYPE(0);
```
Comment (onderkalaci) [inline src/backend/distributed/worker/worker_create_or_replace.c:162]: as far as I know, PG handles the overloaded UDFs by using a different C function and then providing the `link_symbol `, see [here](https://www.postgresql.org/docs/14/sql-createfunction.html) :
```SQL
 CREATE OR REPLACE FUNCTION pg_catalog.worker_create_or_replace_object(statement text) RETURNS bool LANGUAGE C STRICT   AS 'citus', $$worker_create_or_replace_object_text$$;

 CREATE OR REPLACE FUNCTION pg_catalog.worker_create_or_replace_object(statement text[]) RETURNS bool LANGUAGE C STRICT   AS 'citus', $$worker_create_or_replace_object_array$$;
```

If we can use that approach, we can avoid `PG_GETARG_TYPE`

**Review by onderkalaci — APPROVED** [review]
I think it'd be nice to merge this one to the actual PR, and add a test?

Added few minor comments/questions.

File: src/backend/distributed/sql/citus--8.3-1--9.0-1.sql [hunk 3]
```diff
@@ -21,13 +21,7 @@ ALTER FUNCTION citus.restore_isolation_tester_func SET SCHEMA citus_internal;
...
 
 #include "udfs/pg_dist_shard_placement_trigger_func/9.0-1.sql"
-
-CREATE OR REPLACE FUNCTION pg_catalog.worker_create_or_replace_object(statement text)
-  RETURNS bool
-  LANGUAGE C STRICT
-  AS 'MODULE_PATHNAME', $$worker_create_or_replace_object$$;
-COMMENT ON FUNCTION pg_catalog.worker_create_or_replace_object(statement text)
-    IS 'takes a sql CREATE statement, before executing the create it will check if an object with that name already exists and safely replaces that named object with the new object';
+#include "udfs/worker_create_or_replace_object/9.0-1.sql"
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/sql/citus--8.3-1--9.0-1.sql:24]: We don't have to, as long as we are keeping the old function around.
The reason I changed the version from the past was mostly because this UDF wasn't nicely versioned which made it harder to change the function.

If we decide to keep this function now we can do without the 9.0-1 file. However if we decide to _only_ take an array we will require the 9.0-1 file for the downgrade path. If we have the 9.0-1 file I think it is better to use it here as well.

Basically when the function got introduced it should have created this file (and latest.sql in the same dir) to make versioning in the future easier.

File: src/backend/distributed/worker/worker_create_or_replace.c (hunk 2 above)
Comment (thanodnl) [reply src/backend/distributed/worker/worker_create_or_replace.c:162]: I will have a look at that. It will require a bit more refactoring to not duplicate the internal logic too much. Having PG_GETARG_TYPE was a way to work around this. But I think you are right where it might be nicer to have a specialized function for the array variant.

File: src/backend/distributed/sql/citus--8.3-1--9.0-1.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/sql/citus--8.3-1--9.0-1.sql:24]: With using the `link_name` for the binding of C functions to the sql UDFs I have renamed the original C function to reflect the `_text` argument. Hence it is easiest now to have the `9.0-1.sql` file which was missing before.
